### ISSUE TITLE
Still huge offline drain duo2

### ISSUE THREAD

**Issue by Clanceeinfinity** [Archfx/duo-de#130]
### Acknowledgements

- [X] I have chosen an appropriate title.
- [X] All requested information has been provided properly.
- [X] This is not a feature request.
- [X] This issue is not a duplicate of an existing issue.
- [X] The issue is solely related to this GSI.
- [X] This issue occurs with the latest available build.
- [X] I am not running a custom vendor on my device.
- [X] This issue is not kernel related.
- [X] This issue is not related to Magisk/KernelSU/APatch.

### Info

- Device: duo2
- GSI version: v2024.12.16
- Vendor version:

### Expected Behavior

With this update the offline drain should be fixed

### Current Behavior

It's not, the pen charging option is disabled!

### Steps to Reproduce

1. install update 
2. close phone 
3. get drain

### Additional context

![Screenshot_20241217-160913](https://github.com/user-attachments/assets/41ebae08-8438-4d2e-8961-496ffe32a934)
![Screenshot_20241217-160921](https://github.com/user-attachments/assets/5348c0a4-b7ac-44a3-9b0c-dcca9dfa3bc3)

**Comment by Archfx** [comment]
Is this with the wireless pen charger turned off?

**Comment by Clanceeinfinity** [comment]
![Screenshot_20241217-165702](https://github.com/user-attachments/assets/be63ad93-8dcb-4ffa-bd7d-475e7555a10c)
yes, maybe i need to reboot again? it was disabled once i installed the update

**Comment by Archfx** [comment]
Try enabling and disabling, it looks like the default value needs to be set internally. Otherwise, it shows here that it is disabled while internally it is enabled.

I will fix it internally for the next update.

**Comment by Archfx** [comment]
I am closing this issue, feel free to open it again if enabling and disabling doesn't work. I added the internal default value to the sources, so in the next build internally it will be switched off until the manually turned on.

**Comment by Clanceeinfinity** [comment]
tried that too, didn't work :/

**Comment by Archfx** [comment]
Mmm, interesting. Can you get the output of the following commands?

```shell
adb shell
cat /sys/devices/platform/soc/soc:surface_util/ms_pen_charger/ms_pen_charger
```

Does anyone else have the same issue?

**Comment by Clanceeinfinity** [comment]
sadly my pc is still not working, ill test this as soon as it's ready

**Comment by oilcan-productions** [comment]
I am getting permission denied when running the 'cat' command. But not seeing a massive difference in battery drain between pen charging on or off

**Comment by Archfx** [comment]
> I am getting permission denied when running the 'cat' command. But not seeing a massive difference in battery drain between pen charging on or off

If it is getting permission denied, there is a problem. @Ethanol10, I changed the permission for the pen charger file at boot time. But this looks like it is not taking effect.

**Comment by Ethanol10** [comment]
Is the permission set such that user apps can change it too? I can see using ls -Z that the file has `ms_pen_charger` as the policy set for that file. I can only view it if I use su though.

Also, using `ls -l` returns the permission string `-rwxrwxrwx` without changing any permissions, so I'm not entirely convinced this is the problem.

**Comment by Archfx** [comment]
You are correct, `-rwxrwxrwx` means it has the correct permission. I wonder if there are multiple hardware revisions. @oilcan-productions can you get the output for the following commands as well?

```shell
adb shell
ls /sys/devices/platform/soc/soc:surface_util/ms_pen_charger/
ls /sys/devices/platform/soc/soc:surface_util/
ls /sys/devices/platform/soc/

ls -l ls /sys/devices/platform/soc/soc:surface_util/ms_pen_charger/
ls -l /sys/devices/platform/soc/soc:surface_util/
ls -l  /sys/devices/platform/soc/
```

**Comment by oilcan-productions** [comment]
Here you go 
```
duo2:/ $ ls /sys/devices/platform/soc/soc:surface_util/ms_pen_charger/
ls /sys/devices/platform/soc/soc:surface_util/
ls /sys/devices/platfms_pen_charger
....
```

**Comment by Archfx** [comment]
@oilcan-productions thanks for this. Could you please get the output for this command as well?

```shell
adb shell
ls -l /sys/devices/platform/soc/soc:surface_util/ms_pen_charger/ms_pen_charger
```
Do you have access to a pen? Could you please check whether the charger turns on and off with the toggle as well?

**Comment by oilcan-productions** [comment]
```
ls -l /sys/devices/platform/soc/soc:surface_util/ms_pen_charger/ms_pen_charger                                         <
-rw-rw-r-- 1 root root 4096 1970-01-10 05:21 /sys/devices/platform/soc/soc:surface_util/ms_pen_charger/ms_pen_charger
```

Yes, I have a pen and the pen case on my Duo 2. I can confirm that the toggle works. The toggle is off after clean flash and the charger is off as well. When I turn it on the pen starts charging.

**Comment by Archfx** [comment]
> ```
> ls -l /sys/devices/platform/soc/soc:surface_util/ms_pen_charger/ms_pen_charger                                         <
> -rw-rw-r-- 1 root root 4096 1970-01-10 05:21 /sys/devices/platform/soc/soc:surface_util/ms_pen_charger/ms_pen_charger
> ```
> 
> Yes, I have a pen and the pen case on my Duo 2. I can confirm that the toggle works. The toggle is off after clean flash and the charger is off as well. When I turn it on the pen starts charging.

That means the pen charger is working correctly and the unusual battery drain is from another source.

**Comment by oilcan-productions** [comment]
I just clean flashed this morning so not representative right now. I will keep monitoring the battery drain. AS I said I have not noticed a significant difference between pen or no pen charger. I still feel it is VoLTE.iWLAN that chews the battery up.

**Comment by usb4** [comment]
If someone could test that by leaving the Duo2 on airplane mode and seeing if the drain still occurs, that would be a good indicator.

**Comment by oilcan-productions** [comment]
Testing now, will report back once it is charged to 100% and a few hours went in

**Comment by oilcan-productions** [comment]
So far so good. 60 minutes after hitting 100%charge on airplane mode no loss of charge. Pen charger is on pen is paired and docked.

**Comment by oilcan-productions** [comment]
Turned off airplane mode and battery dropped 11% in about an hour. Biggest culprit is Google Play services![Screenshot_20241218-180806.png](https://github.com/user-attachments/assets/a0c28555-87da-494d-a757-51cbb05da1ab)

**Comment by Archfx** [comment]
If Play Services is draining the battery, it should affect both duo1 and duo2 similarly unless it is a problem specific to LTE.

**Comment by Clanceeinfinity** [comment]
play services is not the cause of this, the way android tracks battery is really flawed and if there is an issue where there is abnormal drain, android blames and always active app. With airplane mode enabled, i did lose a little less battery overnight, but still 91 --> 67% so still very abnormal.

**Comment by Ethanol10** [comment]
Does Stock 12L Exhibit the same issues?

e,g airplane mode on, pen charger on etc.

**Comment by oilcan-productions** [comment]
@Ethanol10  I have a12L device sitting right next to the A15 Duo 2 and do not see that high a drain, same configuration with Pen attached, same mobile network.
From the one night test here is what I see. It does seem related to mobile network traffic. Both Devices are on TMO-US, same plan and camping on 5G with Wi-Fi Off. On the 15 device I also turned off the pen charging toggle after removing the pen

| Device | OS  | Start Time | End Time | Start Battery % | End Battery % | AP Mode State | Pen Attached |
|--------|-----|------------|----------|-----------------|---------------|---------------|--------------|
| Duo 2  | 12L | 18:15      | 21:15    | 100             | 99            | ON            | YES          |
| Duo 2  | 15  | 18:17      | 21:30    | 100             | 97            | ON            | YES          |
| Duo 2  | 12L | 21:15      | 00:30    | 99              | 84            | OFF           | YES          |
| Duo 2  | 15  | 21:30      | 00:33    | 97              | 63            | OFF           | YES          |
| Duo 2  | 12L | 00:30      | 08:55    | 84              | 67            | OFF           | NO           |
| Duo 2  | 15  | 00:33      | 08:56    | 63              | 14            | OFF           | NO           |

**Comment by usb4** [comment]
One possible reason that could be causing battery drain is that the fingerprint reader is always active and can unlock the device even while it is folded up.

**Comment by Clanceeinfinity** [comment]
> One possible reason that could be causing battery drain is that the fingerprint reader is always active and can unlock the device even while it is folded up.

i don't think so, most androids work this way. my tab s7 and galaxy s9 / s10 had always active fingerprints too and could go into doze

**Comment by wilhelmstroker** [comment]
When I check the battery usage on mine, Trebuchet is the main culprit.

**Comment by Clanceeinfinity** [comment]
> When I check the battery usage on mine, Trebuchet is the main culprit.

you can not use the battery app to check for abnormal drain sadly. googles sloppy coding doesn't allow for that really. My tab s7 had a similar issue, i think the play services got blamed for abnormal drain, when i checked i saw the cpu being really high used, i think 80% of the drain or so was it (extremely abnormal for the efficient 865+). I noticed that an app was constantly crashing in the background, on this case it was the android telephone subsystem so i deleted it with adb and it got fixed.

main point tho, the real issue, this app, never showed up in the android battery settings app.

**Comment by usb4** [comment]
Another data point:
Went to bed at 100%, woke up to 70%
Airplane mode was on (cellular off), Wifi on
Battery stats says screen was on for 35mins (Duo was folded closed)

Top battery users:
Google Play Store: 
Screen time 11 min
Background 22 min

Trebuchet: 
Screen time 10min
Background **9 hr 2 min**

I turned off "Allow background usage" for Trebuchet, left for a few hours and came back to see that the background usage went up to 14 hrs.

**Comment by usb4** [comment]
> > One possible reason that could be causing battery drain is that the fingerprint reader is always active and can unlock the device even while it is folded up.
> 
> i don't think so, most androids work this way. my tab s7 and galaxy s9 / s10 had always active fingerprints too and could go into doze

I don't believe the stock Duo OS allows you to use the fingerprint reader to unlock the device while it is closed

**Comment by Ethanol10** [comment]
@Archfx I notice in the file `/frameworks/base/core/res/res/values/config.xml`, We do not enable Doze as specified under the variable name: `config_enableAutoPowerModes`, perhaps we should try a build with Doze on to see if we can get any battery improvements from that? Though I have no idea if Doze is being enabled in any of the patches, but from a quick search I haven't found it being enabled in a patch.

https://source.android.com/docs/core/power/platform_mgmt#integrate-doze This config value is mentioned here too, and specifies that it is disabled by default on AOSP, so we could attempt to try a build with this variable integrated, but there might be compatibility issues with system apps, so we must test it before hand.

edit: 
There probably needs to be some optimizations done on top of just enabling the flag in configs,  e,g some people may have notifications coming in late when the device goes into Doze.

**Comment by Archfx** [comment]
> https://source.android.com/docs/core/power/platform_mgmt#integrate-doze

Yep looks like we have to test this. This is from the `config.xml`

```xml
    <!-- Set this to true to enable the platform's auto-power-save modes like doze and
         app standby.  These are not enabled by default because they require a standard
         cloud-to-device messaging service for apps to interact correctly with the modes
         (such as to be able to deliver an instant message to the device even when it is
         dozing).  This should be enabled if you have such services and expect apps to
         correctly use them when installed on your device.  Otherwise, keep this disabled
         so that applications can still use their own mechanisms. -->
    <bool name="config_enableAutoPowerModes">false</bool>
```

**Comment by Archfx** [comment]
Hi @Clanceeinfinity @oilcan-productions, @usb4, @wilhelmstroker we are on the way compiling a DOZE enabled test build.   Will let you guys know when it is ready to test.

**Comment by Archfx** [comment]
DOZE enabled images are available [here](https://github.com/Archfx/duoPosture/releases/tag/v2025.03.06). You guys can test it. 

If this fixes the power consumption related issues, we can release it via OTA.

**Comment by Ethanol10** [comment]
I'll try daily drive it and see how much of a difference there is

**Comment by Clanceeinfinity** [comment]
still doesn't work :(

![Image](https://github.com/user-attachments/assets/ce313557-5e18-4d6d-bb4a-f0a691154353)

**Comment by Ethanol10** [comment]
I'm not sure I understand what you're indicating here, in the photo with the issue we can see the battery drain from 100% to 50% in about 8 hours. In this photo you've shown, it looks like it's going from 80% to 55% in about the same time, 8 hours. 

I understand the battery usage isn't exactly the most efficient, but I do see an improvement.

**Comment by Clanceeinfinity** [comment]
yes its 100% an improvement, but it still seems doze doesn't exactly want to trigger... The cpu is still extremely high so something in the background must be going on

**Comment by Ethanol10** [comment]
Are we talking about the 82% CPU usage during the selected zone? That sounds about right, as the only thing using 82% of the power consumed in that time should be the CPU, as it has to control when apps are able to run in their maintenance windows.

**Comment by Clanceeinfinity** [comment]
i had this issue on my tab s7 too because an app was constantly crashing, cpu was always 80-90% on the battery app. When i uninstalled the app (system telephone) the drain went down from exactly this here to now about 1-2% over night and the battery app reports about 10% cpu usage. Though I can't figure out which app is crashing and i don't want to uninstall the system telephony app because you lose sim support if you do

**Comment by oilcan-productions** [comment]
While the drain with the Doze enabled build is better the change in my testing over the last couple of days is in the vicinity of ~2 additional hours of runtime. Problem is of course if you run continous logging over Wi-Fi to see the events that adds drain as well even with the device closed it will not go to sleep completely as it is connected via adb. but even then it runs around 2 hours longer than a device with the December build

**Comment by rpheuts** [comment]
Not sure this will help, but I had the same issue recompiling the stock 12L kernel after I had backported a couple of syscalls from a newer kernel. I ran out of time trying to get to the bottom of it, but my assumption was that vendor modules were relying on some custom syscalls MS put into the 5.4 kernel that I had to move around. I noticed some crashing behavior in dmesg running my custom kernel. I got the software I was trying to run working with my backported syscalls, but the battery drain was extremely similar to what is being discussed here. I ended up patching out syscalls from libraries instead, as I didn’t have the time to root cause the kernel issue causing the battery drain.

**Comment by Archfx** [comment]
> Not sure this will help, but I had the same issue recompiling the stock 12L kernel after I had backported a couple of syscalls from a newer kernel. I ran out of time trying to get to the bottom of it, but my assumption was that vendor modules were relying on some custom syscalls MS put into the 5.4 kernel that I had to move around. I noticed some crashing behavior in dmesg running my custom kernel. I got the software I was trying to run working with my backported syscalls, but the battery drain was extremely similar to what is being discussed here. I ended up patching out syscalls from libraries instead, as I didn’t have the time to root cause the kernel issue causing the battery drain.

Hi @rpheuts, what kernel version did you try to recompile with? I tried to update the stock kernel to 5.17 or later, and observed that Microsoft changed a lot of things to stock 5.4 to make it work on DUO devices. Actually, it would be cool to get the 6.x version of kernels on the DUO. So if you are interested, we can work on it.
